src/adventure/adventurewidget.cpp

```diff
@@ -0,0 +1,165 @@
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_achievedSomething,
+            this,
+            &AdventureWidget::slot_onAchievedSomething);
+
+    connect(&m_adventureTracker, &AdventureTracker::sig_died, this, &AdventureWidget::slot_onDied);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_gainedLevel,
+            this,
+            &AdventureWidget::slot_onGainedLevel);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_killedMob,
+            this,
+            &AdventureWidget::slot_onKilledMob);
+
+    connect(&m_adventureTracker,
+            &AdventureTracker::sig_receivedHint,
+            this,
+            &AdventureWidget::slot_onReceivedHint);
+}
+
+void AdventureWidget::slot_onAccomplishedTask(double xpGained)
+{
+    // Only record accomplishedTask if it actually has associated xp to avoid
+    // spam, since sometimes it co-triggers with achievemnt and can be redundant.
+    if (xpGained > 0.0) {
+        auto msg = QString(ACCOMPLISH_MSG) + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+        addJournalEntry(msg);
+    }
+}
+
+void AdventureWidget::slot_onAchievedSomething(const QString &achievement, double xpGained)
+{
+    auto msg = QString(ACHIEVE_MSG).arg(achievement);
+
+    if (xpGained > 0.0)
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+
+    addJournalEntry(msg);
+}
+
+void AdventureWidget::slot_onDied(double xpLost)
+{
+    auto msg = QString(DIED_MSG);
+
+    if (xpLost < 0.0)
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpLost));
+
+    addJournalEntry(msg);
+}
+
+void AdventureWidget::slot_onGainedLevel()
+{
+    addJournalEntry(QString(GAINED_LEVEL_MSG));
+}
+
+void AdventureWidget::slot_onKilledMob(const QString &mobName, double xpGained)
+{
+    auto msg = QString(KILL_TROPHY_MSG).arg(mobName);
+
+    if (xpGained > 0.0) {
+        msg = msg + QString(XP_SUFFIX).arg(formatXPGained(xpGained));
+    } else {
+        // When player gets XP from multiple kills on the same heartbeat, like
+        // frequently happens with quake xp, then the first mob gets all the XP
+        // attributed and the others are zero. No way to solve this given
+        // current "protocol" of how MUME exposes information.
+        msg = msg + QString(XP_SUFFIX).arg("?");
+    }
+
+    addJournalEntry(msg);
+}
+
+void AdventureWidget::slot_onReceivedHint(const QString &hint)
+{
+    auto msg = QString(HINT_MSG).arg(hint);
+
+    addJournalEntry(msg);
+}
+
+const QString AdventureWidget::formatXPGained(double xpGained)
+{
+    if (abs(xpGained) < 1000) {
+        return QString::number(xpGained);
+    }
+
+    if (abs(xpGained) < (20 * 1000)) {
+        return QString::number(xpGained / 1000, 'f', 1) + "k";
+    }
+
+    return QString::number(xpGained / 1000, 'f', 0) + "k";
+}
+
+void AdventureWidget::addJournalEntry(const QString &msg)
+{
+    auto prepend = "\n";
```

Comment:
You could write this as `const auto prepend = (m_numMessagesReceived == 0) ? "" : "\n";`

Reply:
fixed